### ISSUE TITLE
Why does calling GetRegistration lock the container?

### ISSUE THREAD

**Issue by kabua** [simpleinjector/SimpleInjector#529]
I have the following configuration:

``` c#
container.RegisterFactory<IUserFactory>();
container.Register(() => new DbContext1("...));
container.GetRegistration(typeof(DbContext)).Registration
    .SuppressDiagnosticWarning(DiagnosticType.DisposableTransientComponent,
        "Called by Abstract Factory");

container.RegisterFactory<IOrgFactory>();
container.Register(() => new DbContext2("..."));
container.GetRegistration(typeof(DbContext2)).Registration
    .SuppressDiagnosticWarning(DiagnosticType.DisposableTransientComponent,
        "Called by Abstract Factory");

container.RegisterWebApiControllers(configuration);
configuration.DependencyResolver = new SimpleInjectorWebApiDependencyResolver(container);

container.Verify();
```

The code fails on RegisterWebApiControllers because the control is locked. Why? I'm still configuring the stuff, I didn't call `GetInstance<T>()`, therefore I think this is or should be a bug.

To get it to work, I had to move the suppression code to the end of the registration process (just before the call to `Verify()`), which doesn't feel natural or right.

**Related Enhancement**

IMHO, I think a better approach would be to have the `Register` functions return the `Registration` object and not have to jump through a bunch of hoops. As in:

``` c#
container.Register(() => new DbContext2("..."))
    .SuppressDiagnosticWarning(DiagnosticType.DisposableTransientComponent, 
    "Called by Abstract Factory");
```
I think this cleaner and more explicit as to what is happening.

**Comment by dotnetjunkie** [comment]
I'm afraid you are correct: this _is_ a bug.

The docentation [describes](https://simpleinjector.readthedocs.io/en/latest/disposabletransientcomponent.html) the use of `GetRegistration` as the way to suppress diagnostic warnings, but unfortunately, a call to `GetRegistration` does lock the container, which makes suppressing warnings very unpractical.

This behavior has been ther

**Comment by dotnetjunkie** [comment]
By the way, is there any particular reason why you wish to register your `DbContext`s as transient? DbContexts should [typically](https://stackoverflow.com/questions/10585478/one-dbcontext-per-web-request-why/10588594#10588594) be reused within a particular scope. You can do this by registering them as `Lifestyle.Scoped`. This ensures them to be disposed automatically by Simple Injector, when the scope ends. This can drastically simplify your application code, since application code isn't responsible anymore of its creation and disposal.

**Comment by dotnetjunkie** [comment]
> IMHO, I think a better approach would be to have the Register functions return the Registration object and not have to jump through a bunch of hoops.

In v5 we plan to introduce a more fluent API, which allows the design you are suggesting. It is planned for v5, since this is a breaking change.

**Comment by dotnetjunkie** [comment]
Fixed with [v4.1.1](https://www.nuget.org/packages/SimpleInjector/4.1.1).
